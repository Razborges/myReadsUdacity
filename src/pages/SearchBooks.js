import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import * as BooksAPI from '../BooksAPI'
import DebounceInput from 'react-debounce-input'
import Book from '../components/Book'

class SearchBooks extends React.Component {
  state = {
    search: '',
    resultBooks: [],
    msg: 'You must enter 3 or more characters to search.'
  }

  handleSearch = (e) => {
    const query = e.target.value
    if (query) {
      this.setState({ search: query })
      this._searchBooks(query)
    } else {
      this.setState({ resultBooks: [], msg: 'You must enter 3 or more characters to search.' })
    }
  }

  _searchBooks = (query) => {
    BooksAPI.search(query, 20)
      .then( resultBooks => {
        if(resultBooks.error) {
          this.setState({ resultBooks: [], msg: 'No results found! You can do a new search.' })
        } else {
          this.props.books.forEach(book => {
            resultBooks.forEach(result => {
              if(book.id === result.id){
                result.shelf = book.shelf
              }
            })
          })
          this.setState({ resultBooks: resultBooks, msg: `Found ${resultBooks.length} books` })
        }
      })
  }

  render() {
    const { resultBooks, msg } = this.state

    return (
      <div className="search-books">
        <div className="search-books-bar">
          <Link className="close-search"  to="/">Close</Link>
          <div className="search-books-input-wrapper">
            <DebounceInput
              minLength={3}
              debounceTimeout={200}
              element="input"
              type="text"
              placeholder="Search by title or author"
              onChange={this.handleSearch}
            />
          </div>
        </div>
        <div className="search-books-results">
          <div>
            <div>
              <h3>{ msg }</h3>
            </div>
            { resultBooks.length > 0 && (
            <ol className="books-grid">
              {resultBooks.map((book) => (
                <Book
                  book={ book }
                  key={ book.id }
                  changeBook={ this.props.changeBook }
                />
              ))}
            </ol>
            )}
          </div>
        </div>
      </div>
    )
  }

  static proptypes = {
    books: PropTypes.array.isRequired,
    changeBook: PropTypes.func.isRequired
  }
}

export default SearchBooks
