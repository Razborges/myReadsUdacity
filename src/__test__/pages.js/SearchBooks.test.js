import React from 'react';
import { shallow, mount } from 'enzyme';
import SearchBooks from './../../pages/SearchBooks';
import { MemoryRouter } from 'react-router-dom';

import booksMock from './../__mock__/booksMock';

describe('<SearchBooks />', () => {
  const wrapper = mount(
    <MemoryRouter>
      <SearchBooks
        books={ booksMock.books }
        changeBook={ jest.fn() }
      />
    </MemoryRouter>)

  afterEach(() => wrapper.setState({
    search: '',
    resultBooks: [],
    msg: 'You must enter 3 or more characters to search.'
  }));

  it('shallow renders without crashing', () => {
    expect(shallow(<SearchBooks books={ booksMock } changeBook={ jest.fn() } />))
  });

  it('snapshot SearchBooks', () => {
    expect(shallow(<SearchBooks books={ booksMock } changeBook={ jest.fn() } />)).toMatchSnapshot();
  });

  it('initial state search query to empty', () => {
    expect(wrapper.state().search.length).toBe(0)
  });

  it('initial state resultBooks to empty', () => {
    expect(wrapper.state().resultBooks.length).toBe(0)
  });

  it('has DebounceInput component', () => {
    expect(wrapper.find('DebounceInput').length).toBe(1)
  });
})
