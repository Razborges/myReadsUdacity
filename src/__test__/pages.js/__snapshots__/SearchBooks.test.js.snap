exports[`<SearchBooks /> snapshot SearchBooks 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchBooks
    books={
        Object {
            "books": Array [
              Object {
                "allowAnonLogging": true,
                "authors": Array [
                  "William E. Shotts, Jr.",
                ],
                "averageRating": 4,
                "canonicalVolumeLink": "https://market.android.com/details?id=book-nggnmAEACAAJ",
                "categories": Array [
                  "COMPUTERS",
                ],
                "contentVersion": "1.2.2.0.preview.2",
                "description": "You\'ve experienced the shiny, point-and-click surface of your Linux computer—now dive below and explore its depths with the power of the command line. The Linux Command Line takes you from your very first terminal keystrokes to writing full programs in Bash, the most popular Linux shell. Along the way you\'ll learn the timeless skills handed down by generations of gray-bearded, mouse-shunning gurus: file navigation, environment configuration, command chaining, pattern matching with regular expressions, and more. In addition to that practical knowledge, author William Shotts reveals the philosophy behind these tools and the rich heritage that your desktop Linux machine has inherited from Unix supercomputers of yore. As you make your way through the book\'s short, easily-digestible chapters, you\'ll learn how to: * Create and delete files, directories, and symlinks * Administer your system, including networking, package installation, and process management * Use standard input and output, redirection, and pipelines * Edit files with Vi, the world’s most popular text editor * Write shell scripts to automate common or boring tasks * Slice and dice text files with cut, paste, grep, patch, and sed Once you overcome your initial \"shell shock,\" you\'ll find that the command line is a natural and expressive way to communicate with your computer. Just don\'t be surprised if your mouse starts to gather dust. A featured resource in the Linux Foundation\'s \"Evolution of a SysAdmin\"",
                "id": "nggnmAEACAAJ",
                "imageLinks": Object {
                  "smallThumbnail": "http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=5&source=gbs_api",
                  "thumbnail": "http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api",
                },
                "industryIdentifiers": Array [
                  Object {
                    "identifier": "9781593273897",
                    "type": "ISBN_13",
                  },
                  Object {
                    "identifier": "1593273894",
                    "type": "ISBN_10",
                  },
                ],
                "infoLink": "https://play.google.com/store/books/details?id=nggnmAEACAAJ&source=gbs_api",
                "language": "en",
                "maturityRating": "NOT_MATURE",
                "pageCount": 480,
                "panelizationSummary": Object {
                  "containsEpubBubbles": false,
                  "containsImageBubbles": false,
                },
                "previewLink": "http://books.google.com/books?id=nggnmAEACAAJ&dq=linux&hl=&cd=3&source=gbs_api",
                "printType": "BOOK",
                "publishedDate": "2012",
                "publisher": "No Starch Press",
                "ratingsCount": 2,
                "readingModes": Object {
                  "image": false,
                  "text": true,
                },
                "shelf": "read",
                "subtitle": "A Complete Introduction",
                "title": "The Linux Command Line",
              },
              Object {
                "allowAnonLogging": false,
                "authors": Array [
                  "Harmeet Singh",
                  "Mehul Bhatt",
                ],
                "canonicalVolumeLink": "https://books.google.com/books/about/Learning_Web_Development_with_React_and.html?hl=&id=sJf1vQAACAAJ",
                "contentVersion": "preview-1.0.0",
                "description": "Build real-time responsive web apps using React and BootstrapAbout This Book* Showcase the power of React-Bootstrap through real-world examples* Explore the benefits of integrating React with various frameworks and APIs* See the benefits of using the latest frameworks to make your web development experience enchantingWho This Book Is ForThis book is for anybody who is interested in modern web development and has intermediate knowledge of HTML, CSS, and JavaScript. Basic knowledge of any JavaScript MVC framework would also be helpful.What You Will Learn* See how to integrate React-Bootstrap with React* Explore the Redux architecture and understand its benefits* Build a custom responsive theme* Easily interact with DOM on your web browser* Appreciate the advantages of using JSX* Get acquainted with the various routing methods in React* Integrate external APIs into ReactIn DetailReact-Bootstrap is one of the most popular front-end frameworks, and integrating it with React allows web developers to write much cleaner code. This book will help you gain a thorough understanding of the React-Bootstrap framework and show you how to build impressive web apps.In this book, you will get an overview of the features of React-Bootstrap and ReactJS, along with the integration of ReactJS components with ReactJS. You will understand the benefits of using JSX and the Redux architecture. The server-side rendering of React will also be shown. All the concepts are explained by developing real-world examples.By the end of this book, you will be equipped to create responsive web applications using React-Bootstrap with ReactJS, and will have an insight into the best practices.",
                "id": "sJf1vQAACAAJ",
                "imageLinks": Object {
                  "smallThumbnail": "http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=5&source=gbs_api",
                  "thumbnail": "http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api",
                },
                "industryIdentifiers": Array [
                  Object {
                    "identifier": "1786462494",
                    "type": "ISBN_10",
                  },
                  Object {
                    "identifier": "9781786462497",
                    "type": "ISBN_13",
                  },
                ],
                "infoLink": "http://books.google.com/books?id=sJf1vQAACAAJ&dq=redux+react&hl=&source=gbs_api",
                "language": "en",
                "maturityRating": "NOT_MATURE",
                "pageCount": 278,
                "panelizationSummary": Object {
                  "containsEpubBubbles": false,
                  "containsImageBubbles": false,
                },
                "previewLink": "http://books.google.com/books?id=sJf1vQAACAAJ&dq=redux+react&hl=&cd=6&source=gbs_api",
                "printType": "BOOK",
                "publishedDate": "2016-12-30",
                "readingModes": Object {
                  "image": false,
                  "text": false,
                },
                "shelf": "currentlyReading",
                "title": "Learning Web Development with React and Bootstrap",
              },
              Object {
                "allowAnonLogging": true,
                "authors": Array [
                  "Robert Galbraith",
                ],
                "averageRating": 3.5,
                "canonicalVolumeLink": "https://market.android.com/details?id=book-evuwdDLfAyYC",
                "categories": Array [
                  "Fiction",
                ],
                "contentVersion": "1.21.20.0.preview.2",
                "description": "A brilliant debut mystery in a classic vein: Detective Cormoran Strike investigates a supermodel\'s suicide. After losing his leg to a land mine in Afghanistan, Cormoran Strike is barely scraping by as a private investigator. Strike is down to one client, and creditors are calling. He has also just broken up with his longtime girlfriend and is living in his office. Then John Bristow walks through his door with an amazing story: His sister, thelegendary supermodel Lula Landry, known to her friends as the Cuckoo, famously fell to her death a few months earlier. The police ruled it a suicide, but John refuses to believe that. The case plunges Strike into the world of multimillionaire beauties, rock-star boyfriends, and desperate designers, and it introduces him to every variety of pleasure, enticement, seduction, and delusion known to man. You may think you know detectives, but you\'ve never met one quite like Strike. You may think you know about the wealthy and famous, but you\'ve never seen them under an investigation like this. Introducing Cormoran Strike, this is the acclaimed first crime novel by J.K. Rowling, writing under the pseudonym Robert Galbraith.",
                "id": "evuwdDLfAyYC",
                "imageLinks": Object {
                  "smallThumbnail": "http://books.google.com/books/content?id=evuwdDLfAyYC&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api",
                  "thumbnail": "http://books.google.com/books/content?id=evuwdDLfAyYC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
                },
                "industryIdentifiers": Array [
                  Object {
                    "identifier": "9780316206860",
                    "type": "ISBN_13",
                  },
                  Object {
                    "identifier": "0316206865",
                    "type": "ISBN_10",
                  },
                ],
                "infoLink": "https://play.google.com/store/books/details?id=evuwdDLfAyYC&source=gbs_api",
                "language": "en",
                "maturityRating": "NOT_MATURE",
                "pageCount": 464,
                "panelizationSummary": Object {
                  "containsEpubBubbles": false,
                  "containsImageBubbles": false,
                },
                "previewLink": "http://books.google.com/books?id=evuwdDLfAyYC&printsec=frontcover&dq=rowling&hl=&cd=4&source=gbs_api",
                "printType": "BOOK",
                "publishedDate": "2013-04-30",
                "publisher": "Mulholland Books",
                "ratingsCount": 3037,
                "readingModes": Object {
                  "image": false,
                  "text": true,
                },
                "shelf": "wantToRead",
                "title": "The Cuckoo\'s Calling",
              },
            ],
          }
    }
    changeBook={[Function]} />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="search-books-bar">
          <Link
                    className="close-search"
                    replace={false}
                    to="/">
                    Close
          </Link>
          <div
                    className="search-books-input-wrapper">
                    <DebounceInput
                              debounceTimeout={200}
                              element="input"
                              forceNotifyByEnter={true}
                              forceNotifyOnBlur={true}
                              inputRef={undefined}
                              minLength={3}
                              onBlur={undefined}
                              onChange={[Function]}
                              onKeyDown={undefined}
                              placeholder="Search by title or author"
                              type="text"
                              value={undefined} />
          </div>
</div>,
        <div
          className="search-books-results">
          <div>
                    <div>
                              <h3>
                                        You must enter 3 or more characters to search.
                              </h3>
                    </div>
          </div>
</div>,
      ],
      "className": "search-books",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Link
              className="close-search"
              replace={false}
              to="/">
              Close
</Link>,
            <div
              className="search-books-input-wrapper">
              <DebounceInput
                            debounceTimeout={200}
                            element="input"
                            forceNotifyByEnter={true}
                            forceNotifyOnBlur={true}
                            inputRef={undefined}
                            minLength={3}
                            onBlur={undefined}
                            onChange={[Function]}
                            onKeyDown={undefined}
                            placeholder="Search by title or author"
                            type="text"
                            value={undefined} />
</div>,
          ],
          "className": "search-books-bar",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Close",
              "className": "close-search",
              "replace": false,
              "to": "/",
            },
            "ref": null,
            "rendered": "Close",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <DebounceInput
                debounceTimeout={200}
                element="input"
                forceNotifyByEnter={true}
                forceNotifyOnBlur={true}
                inputRef={undefined}
                minLength={3}
                onBlur={undefined}
                onChange={[Function]}
                onKeyDown={undefined}
                placeholder="Search by title or author"
                type="text"
                value={undefined} />,
              "className": "search-books-input-wrapper",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "debounceTimeout": 200,
                "element": "input",
                "forceNotifyByEnter": true,
                "forceNotifyOnBlur": true,
                "inputRef": undefined,
                "minLength": 3,
                "onBlur": undefined,
                "onChange": [Function],
                "onKeyDown": undefined,
                "placeholder": "Search by title or author",
                "type": "text",
                "value": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div>
            <div>
                        <h3>
                                    You must enter 3 or more characters to search.
                        </h3>
            </div>
</div>,
          "className": "search-books-results",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div>
                <h3>
                                You must enter 3 or more characters to search.
                </h3>
</div>,
              false,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <h3>
                  You must enter 3 or more characters to search.
</h3>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "You must enter 3 or more characters to search.",
                },
                "ref": null,
                "rendered": "You must enter 3 or more characters to search.",
                "type": "h3",
              },
              "type": "div",
            },
            false,
          ],
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="search-books-bar">
            <Link
                        className="close-search"
                        replace={false}
                        to="/">
                        Close
            </Link>
            <div
                        className="search-books-input-wrapper">
                        <DebounceInput
                                    debounceTimeout={200}
                                    element="input"
                                    forceNotifyByEnter={true}
                                    forceNotifyOnBlur={true}
                                    inputRef={undefined}
                                    minLength={3}
                                    onBlur={undefined}
                                    onChange={[Function]}
                                    onKeyDown={undefined}
                                    placeholder="Search by title or author"
                                    type="text"
                                    value={undefined} />
            </div>
</div>,
          <div
            className="search-books-results">
            <div>
                        <div>
                                    <h3>
                                                You must enter 3 or more characters to search.
                                    </h3>
                        </div>
            </div>
</div>,
        ],
        "className": "search-books",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Link
                className="close-search"
                replace={false}
                to="/">
                Close
</Link>,
              <div
                className="search-books-input-wrapper">
                <DebounceInput
                                debounceTimeout={200}
                                element="input"
                                forceNotifyByEnter={true}
                                forceNotifyOnBlur={true}
                                inputRef={undefined}
                                minLength={3}
                                onBlur={undefined}
                                onChange={[Function]}
                                onKeyDown={undefined}
                                placeholder="Search by title or author"
                                type="text"
                                value={undefined} />
</div>,
            ],
            "className": "search-books-bar",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Close",
                "className": "close-search",
                "replace": false,
                "to": "/",
              },
              "ref": null,
              "rendered": "Close",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <DebounceInput
                  debounceTimeout={200}
                  element="input"
                  forceNotifyByEnter={true}
                  forceNotifyOnBlur={true}
                  inputRef={undefined}
                  minLength={3}
                  onBlur={undefined}
                  onChange={[Function]}
                  onKeyDown={undefined}
                  placeholder="Search by title or author"
                  type="text"
                  value={undefined} />,
                "className": "search-books-input-wrapper",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "debounceTimeout": 200,
                  "element": "input",
                  "forceNotifyByEnter": true,
                  "forceNotifyOnBlur": true,
                  "inputRef": undefined,
                  "minLength": 3,
                  "onBlur": undefined,
                  "onChange": [Function],
                  "onKeyDown": undefined,
                  "placeholder": "Search by title or author",
                  "type": "text",
                  "value": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div>
              <div>
                            <h3>
                                          You must enter 3 or more characters to search.
                            </h3>
              </div>
</div>,
            "className": "search-books-results",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div>
                  <h3>
                                    You must enter 3 or more characters to search.
                  </h3>
</div>,
                false,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <h3>
                    You must enter 3 or more characters to search.
</h3>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "You must enter 3 or more characters to search.",
                  },
                  "ref": null,
                  "rendered": "You must enter 3 or more characters to search.",
                  "type": "h3",
                },
                "type": "div",
              },
              false,
            ],
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
